core/types/l2trace.go

```diff
@@ -1,34 +1,92 @@
 package types
 
 import (
+	"encoding/json"
+	"math/big"
+
 	"github.com/scroll-tech/go-ethereum/common"
 	"github.com/scroll-tech/go-ethereum/common/hexutil"
 )
 
 // BlockResult contains block execution traces and results required for rollers.
 type BlockResult struct {
 	BlockTrace       *BlockTrace        `json:"blockTrace"`
+	StorageTrace     *StorageTrace      `json:"storageTrace"`
 	ExecutionResults []*ExecutionResult `json:"executionResults"`
 }
 
+// StorageTrace stores proofs of storage needed by storage circuit
+type StorageTrace struct {
+	// Root hash before block execution:
+	RootBefore common.Hash `json:"rootBefore,omitempty"`
+	// Root hash after block execution, is nil if execution has failed
+	RootAfter common.Hash `json:"rootAfter,omitempty"`
+
+	// All proofs BEFORE execution, for accounts which would be used in tracing
+	Proofs map[string][]hexutil.Bytes `json:"proofs"`
+
+	// All storage proofs BEFORE execution
+	StorageProofs map[string]map[string][]hexutil.Bytes `json:"storageProofs,omitempty"`
+}
+
+// EvmTxTraces groups trace data from each executation of tx and left
+// some field to be finished from blockResult
+type EvmTxTraces struct {
+	TxResults []*ExecutionResult
+	Storage   *StorageTrace
+}
+
 // ExecutionResult groups all structured logs emitted by the EVM
 // while replaying a transaction in debug mode as well as transaction
 // execution status, the amount of gas used and the return value
 type ExecutionResult struct {
 	Gas         uint64 `json:"gas"`
 	Failed      bool   `json:"failed"`
 	ReturnValue string `json:"returnValue,omitempty"`
-	// Sender's account proof.
-	From *AccountProofWrapper `json:"from,omitempty"`
+	// Sender's account proof (before Tx)..
+	From *AccountWrapper `json:"from,omitempty"`
 	// Receiver's account proof.
-	To *AccountProofWrapper `json:"to,omitempty"`
+	To *AccountWrapper `json:"to,omitempty"`
+	// AccountCreated record the account in case tx is create
+	// (for creating inside contracts we handle CREATE op)
+	AccountCreated *AccountWrapper `json:"accountCreated,omitempty"`
+
+	// Record all accounts' state which would be affected AFTER tx executed
+	// currently they are just sender and to account
+	AccountsAfter []*AccountWrapper `json:"accountAfter"`
+
 	// It's exist only when tx is a contract call.
 	CodeHash *common.Hash `json:"codeHash,omitempty"`
 	// If it is a contract call, the contract code is returned.
-	ByteCode   string         `json:"byteCode,omitempty"`
+	ByteCode string `json:"byteCode,omitempty"`
+
 	StructLogs []StructLogRes `json:"structLogs"`
 }
 
+// HexInt wrap big.Int for hex encoding
+type HexInt struct {
```

Comment:
How about moving to the Common module？

Reply:
to hexutil module may be better?

Reply:
> to hexutil module may be better?

I agree, or just reuse hexutil.Big?

Reply:
That's good, i will just purged it because it is just used in the SMTTrace output (not included in current PR) and use hexutil.Big instead later